src/kv/RocksDBStore.h

```diff
@@ -341,9 +353,29 @@ class RocksDBStore : public KeyValueDB {
     public KeyValueDB::WholeSpaceIteratorImpl {
   protected:
     rocksdb::Iterator *dbiter;
+    const KeyValueDB::IteratorBounds bounds;
+    const rocksdb::Slice iterate_lower_bound;
+    const rocksdb::Slice iterate_upper_bound;
   public:
-    explicit RocksDBWholeSpaceIteratorImpl(rocksdb::Iterator *iter) :
-      dbiter(iter) { }
+    explicit RocksDBWholeSpaceIteratorImpl(const RocksDBStore* db,
+                                           rocksdb::ColumnFamilyHandle* cf,
+                                           const KeyValueDB::IteratorOpts opts,
+                                           const KeyValueDB::IteratorBounds bounds_) :
+      bounds(std::move(bounds_)),
```

Comment:
note that move semantics require non-const input - we want `bounds` to 'steal' the memory owned by `bounds_`, and it can't do that safely without modifying `bounds_` - otherwise its destructor would free the memory that `bounds` is still referencing

another way to look at it, is that we're trying to call the compiler-generated move constructor of `IteratorBounds`:
```c++
  IteratorBounds(IteratorBounds&& other) = default; // move constructor
  IteratorBounds(const IteratorBounds& other) = default; // copy constructor
```
`std::move()` is just a cast to rvalue-reference `&&`, so given `const IteratorBounds bounds_`, the return type of `std::move(bounds_)` would be `const IteratorBounds&&`. this doesn't match the move constructor overload, so would call the copy constructor instead

tldr: remove the `const` from `bounds_` here, and do the same for `ShardMergeIteratorImpl` and `CFIteratorImpl`

Reply:
excellent explanation, thanks again @cbodley !